comms/storage/decider/RendezvousDecider.go

```diff
@@ -82,12 +87,51 @@ func (d *RendezvousDecider) computeShardsNodeStores(publicKey string) []string {
 	return shards
 }
 
-// getHashRing returns a data structure that spreads storage across all allStorageNodePubKeys in the network.
-// N.b. this implementation currently uses rendezvous/highest random weight hashing so it's not technically a ring.
-func (d *RendezvousDecider) getHashRing() *rendezvous.Hash {
+// getRendezvousHash returns a data structure that spreads storage across all allStorageNodePubKeys in the network.
+// Uses rendezvous/highest random weight hashing.
+func (d *RendezvousDecider) getRendezvousHash() *rendezvous.Hash {
 	hash := rendezvous.New()
-	for _, publicKey := range d.allStorageNodePubKeys {
+	for _, publicKey := range d.healthyNodePubKeys {
 		hash.Add(publicKey)
 	}
 	return hash
 }
+
+func (d *RendezvousDecider) watchForNewHealthyNodeSet(kv nats.KeyValue) {
+	watcher, err := kv.Watch(d.namespace)
+	if err != nil {
+		log.Fatal("failed to watch for changes to set of healthy nodes: ", err)
+	}
+
+	go func() {
+		for change := range watcher.Updates() {
+			if change == nil {
+				continue
+			}
+
+			// Read latest set of healthy nodes
+			var healthyNodePubKeys []string
+			err := json.Unmarshal(change.Value(), &healthyNodePubKeys)
+			if err != nil {
+				log.Print("error: invalid healthyNodes KV value: " + string(change.Value()))
+				continue
+			}
+
+			// Update this node's view of the set of healthy nodes to match the latest
+			prevShards := d.ShardsStored
+			d.healthyNodePubKeys = healthyNodePubKeys
+			d.ShardsStored = d.computeShardsNodeStores()
+			curShards := d.ShardsStored
+			rebalanceShardsStored(prevShards, curShards)
+		}
+	}()
+}
+
+// rebalanaceShardsStored finds content that needs to be stored or deleted and fetches or deletes it.
+func rebalanceShardsStored(prevShards []string, curShards []string) error {
```

Comment:
this will also need to
1. bypass premium content checks in a way that only registered nodes can
2. handle blacklisted content correctly.
you maybe tracking these already

Reply:
good callout, thanks - added a note [here](https://linear.app/audius/issue/CON-582/implement-content-rebalancing) to address those when implementing rebalancing